server/src/test/scala/services/graphstages/ArrivalsGraphStageSpec.scala

```diff
@@ -0,0 +1,178 @@
+    import akka.stream.scaladsl.GraphDSL.Implicits._
+
+    val graph = GraphDSL.create(
+      liveArrivalsSource.async,
+      liveBaseArrivalsSource.async,
+      forecastArrivalsSource.async,
+      forecastBaseArrivalsSource.async
+    )((_, _, _, _)) {
+
+      implicit builder =>
+        (live, liveBase, forecast, forecastBase) =>
+
+          val arrivals = builder.add(arrivalsGraphStage.async)
+          val sink = builder.add(Sink.actorRef(testProbe.ref, "complete"))
+
+          forecastBase ~> arrivals.in0
+          forecast ~> arrivals.in1
+          liveBase ~> arrivals.in2
+          live ~> arrivals.in3
+          arrivals.out ~> sink
+
+          ClosedShape
+    }
+
+    RunnableGraph.fromGraph(graph)
+  }
+}
+
+class ArrivalsGraphStageSpec extends CrunchTestLike {
+
+  "Given a live arrival and a base live arrival I should get a merged arrival" >> {
+    val probe = TestProbe("arrivals")
+    val arrivalsGraphStage = new ArrivalsGraphStage("",
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      a => MilliDate(a.Scheduled),
+      Set("T1"),
+      Crunch.oneDayMillis,
+      () => SDate(2019, 10, 1, 16, 0)
+    )
+    val (liveSource, liveBaseSource, _, _) = TestableArrivalsGraphStage(probe, arrivalsGraphStage).run
+    val liveEstimated = Option(SDate(2019, 10, 1, 16, 0).millisSinceEpoch)
+    val liveBaseActual = Option(SDate(2019, 10, 1, 16, 0).millisSinceEpoch)
+
+    liveSource.offer(ArrivalsFeedSuccess(Flights(List(arrival(estimated = liveBaseActual)))))
+    liveBaseSource.offer(ArrivalsFeedSuccess(Flights(List(arrival(actual = liveBaseActual)))))
+
+    probe.fishForMessage(10 seconds) {
+      case ArrivalsDiff(toUpdate, _) =>
+        toUpdate.exists {
+          case (_, a) => a.Estimated == liveEstimated && a.Actual == liveBaseActual
+        }
+    }
+
+    success
+  }
+
+  "Given a base live arrival only we should not get anything back" >> {
+    val probe = TestProbe("arrivals")
+    val arrivalsGraphStage = new ArrivalsGraphStage("",
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      mutable.SortedMap[UniqueArrival, Arrival](),
+      a => MilliDate(a.Scheduled),
+      Set("T1"),
+      Crunch.oneDayMillis,
+      () => SDate(2019, 10, 1, 16, 0)
+    )
+    val (_, liveBaseSource, _, _) = TestableArrivalsGraphStage(probe, arrivalsGraphStage).run
+    val liveBaseActual = Option(SDate(2019, 10, 1, 16, 0).millisSinceEpoch)
+
+    liveBaseSource.offer(ArrivalsFeedSuccess(Flights(List(arrival(actual = liveBaseActual)))))
+
+    probe.fishForMessage(10 seconds) {
```

Comment:
prob safe to reduce this to the region of 2 or 3 seconds

Reply:
could possibly replace it with explicitly checking we don't receive a message containing flights in case we happen to hit on a valid empty message but a non-empty message follows soon after